### PR TITLE
Rewrite normalize action

### PR DESCRIPTION
**Pull request by nahueespinosa**
### Proposed changes

Trying a new way of implementing actions that gets rid of the argument re-ordering madness and is hopefully easier to maintain.

#### Type of change

- [x] 🐛 Bugfix (change which fixes an issue)
- [ ] 🚀 Feature (change which adds functionality)
- [ ] 📚 Documentation (change which fixes or extends documentation)

### Checklist

- [x] Lint and unit tests (if any) pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)
- [x] All commits have been signed for [DCO](https://developercertificate.org/)

### CODE REVIEW

**Review by hidmic — COMMENTED** [review]
Overall LGTM. The one thing I don't love is the API break. I'd rather define and immediately deprecate pipe operators to keep the no argument variant operational instead of dropping.

**Comment by nahueespinosa** [comment]
> The one thing I don't love is the API break

Fixed in https://github.com/Ekumen-OS/beluga/pull/441/commits/185f744f89d142cfe1c13450b63647bb537d1fbb.

File: beluga/include/beluga/actions/normalize.hpp [hunk 1]
```diff
@@ -34,120 +35,106 @@ namespace beluga::actions {
...
-    }();
+    });
+
+    const double factor = std::invoke([this, weights]() {
+      if (factor_.has_value()) {
+        return factor_.value();
+      }
+
+      return ranges::accumulate(weights, 0.0);  // The default normalization factor is the total sum of weights.
+    });
```
Comment (DPR00) [inline beluga/include/beluga/actions/normalize.hpp:71]: Could this be replaced by:
```cpp
const double factor = factor_.value_or(ranges::accumulate(weights, 0.0));
```
?

File: beluga/include/beluga/actions/normalize.hpp [hunk 2]
```diff
@@ -34,120 +35,106 @@ namespace beluga::actions {
...
     std::transform(
-        policy,               //
-        std::begin(weights),  //
-        std::end(weights),    //
-        std::begin(weights),  //
+        policy_,          //
+        weights.begin(),  //
+        weights.end(),    //
+        weights.begin(),  //
         [factor](const auto w) { return w / factor; });
```
Comment (DPR00) [inline beluga/include/beluga/actions/normalize.hpp:82]: meta: Should we add some kind of `assert` to ensure that factor won't take a zero value? Probably when weights is empty.

**Review by DPR00 — COMMENTED** [review]
Awesome! All the actions should take this structure, right? If so, I can help with the overlay action.

File: beluga/include/beluga/actions/normalize.hpp (hunk 1 above)
Comment (nahueespinosa) [reply beluga/include/beluga/actions/normalize.hpp:71]: That would accumulate the weights every time, and then decide whether to use a user-provided factor or the sum. This version only accumulates weights if there is no user-provided factor.

I'd like to use C++20's monadic operations, but this is C++17. See [`or_else`](https://en.cppreference.com/w/cpp/utility/optional/or_else).

**Review by hidmic — APPROVED** [review]

**Comment by nahueespinosa** [comment]
> All the actions should take this structure, right? If so, I can help with the overlay action.

Yeah, that would be great!

**Pull request merged**